File: 2009/S092.cpp
#include <cstdio>
int gifts[500000];
int last[1001];
int w[50000], v[50000];
int main() {
  int n, m;
  scanf("%d %d", &n, &m);
  for (int i = 0; i < n; i++) {
    scanf("%d", &gifts[i]);
  }
  for (int i = 0; i < m; i++) {
    scanf("%d %d", &w[i], &v[i]);
  }
  for (int i = 0; i < n; i++) {
    while (last[gifts[i]] < m && (w[last[gifts[i]]] < gifts[i] || v[last[gifts[i]]] < gifts[i])) {
      last[gifts[i]]++;
    }
    if (last[gifts[i]] == m) {
      printf("Impossible\n");
      return 0;
    }
    v[last[gifts[i]]] -= gifts[i];
    gifts[i] = last[gifts[i]] + 1;
  }
  for (int i = 0; i < n - 1; i++) {
    printf("%d ", gifts[i]);
  }
  printf("%d\n", gifts[n - 1]);
  return 0;
}

File: 2009/S092.pas
var
  gifts: array[1..500000] of longint;
  last: array[1..1000] of longint;
  w, v: array[1..50000] of longint;
  n, m, i: longint;
begin
  for i := 1 to 1000 do
    last[i] := 1;
  read(n, m);
  for i := 1 to n do
    read(gifts[i]);
  for i := 1 to m do
    read(w[i], v[i]);
  for i := 1 to n do
  begin
    while ((last[gifts[i]] <= m) and ((w[last[gifts[i]]] < gifts[i]) or (v[last[gifts[i]]] < gifts[i]))) do
      inc(last[gifts[i]]);
    if (last[gifts[i]] = m + 1) then
    begin
      writeln('Impossible');
      halt
    end;
    v[last[gifts[i]]] := v[last[gifts[i]]] - gifts[i];
    gifts[i] := last[gifts[i]]
  end;
  for i := 1 to n - 1 do
    write(gifts[i], ' ');
  writeln(gifts[n])
end.
